### PR TITLE
minor fix for customized view aggregation

### PR DESCRIPTION
**Pull request by zhangmeng916**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

(#885 )

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:
This PR fixed a few of minor issues in customized aggregation view and add one more test. The main change is to change lastSeenCustomizedStateType to a map and a few logic change that handle corner cases.

### Tests

- [X] The following tests are written for this issue:
TestComputeAndCleanupCustomizedView

- [X] The following is the result of the "mvn test" command on the appropriate module:

### Commits

- [X ] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [X] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java [hunk 1]
```diff
@@ -139,7 +139,7 @@
   final AtomicReference<Map<String, LiveInstance>> _lastSeenInstances;
   final AtomicReference<Map<String, LiveInstance>> _lastSeenSessions;
 
-  final AtomicReference<Set<String>> _lastSeenCustomizedStateTypes;
+  final AtomicReference<Map<String, Set<String>>> _lastSeenCustomizedStateTypesMap;
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: More comments about the content of the map, please.

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java [hunk 2]
```diff
@@ -828,8 +828,14 @@ public void onCustomizedStateRootChange(String instanceName, List<String> custom
     }
 
     // TODO: remove the synchronization here once we move this update into dataCache.
-    synchronized (_lastSeenCustomizedStateTypes) {
-      Set<String> lastSeenCustomizedStateTypes = _lastSeenCustomizedStateTypes.get();
+    synchronized (_lastSeenCustomizedStateTypesMap) {
+      Map<String, Set<String>> lastSeenCustomizedStateTypesMap =
+          _lastSeenCustomizedStateTypesMap.get();
+      Set<String> lastSeenCustomizedStateTypes = new HashSet<>();
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: nit Collections.emtpySet()

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java [hunk 3]
```diff
@@ -828,8 +828,14 @@ public void onCustomizedStateRootChange(String instanceName, List<String> custom
...
+      if (lastSeenCustomizedStateTypesMap != null && lastSeenCustomizedStateTypesMap
+          .containsKey(instanceName)) {
+        lastSeenCustomizedStateTypes = lastSeenCustomizedStateTypesMap.get(instanceName);
+      }
       for (String customizedState : customizedStateTypes) {
         try {
           if (lastSeenCustomizedStateTypes == null || !lastSeenCustomizedStateTypes
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: lastSeenCustomizedStateTypes won't be null, right?

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java [hunk 4]
```diff
@@ -844,14 +850,24 @@ public void onCustomizedStateRootChange(String instanceName, List<String> custom
         }
       }
 
-      for (String previousCustomizedState : lastSeenCustomizedStateTypes) {
-        if (!customizedStateTypes.contains(previousCustomizedState)) {
-          manager.removeListener(keyBuilder.customizedStates(instanceName, previousCustomizedState),
-              this);
+      if (lastSeenCustomizedStateTypes != null) {
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: Same here, do we really need to check this?

File: helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java [hunk 5]
```diff
@@ -197,18 +197,17 @@ private void refreshTargetExternalViews(final HelixDataAccessor accessor) {
   public void refreshCustomizedViewMap(final HelixDataAccessor accessor) {
     // As we are not listening on customized view change, customized view will be
     // refreshed once during the cache's first refresh() call, or when full refresh is required
-    List<String> newStateTypes = accessor.getChildNames(accessor.keyBuilder().customizedViews());
     if (_propertyDataChangedMap.get(HelixConstants.ChangeType.CUSTOMIZED_VIEW).getAndSet(false)) {
-      for (String stateType : newStateTypes) {
+      for (String stateType : _aggregationEnabledTypes) {
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java]: This does not look right. If the config types have been changed just before the refresh with fewer items, we won't remove the extra types, right?

File: helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java [hunk 6]
```diff
@@ -197,18 +197,17 @@ private void refreshTargetExternalViews(final HelixDataAccessor accessor) {
...
           _customizedViewCacheMap.put(stateType, newCustomizedViewCache);
         }
         _customizedViewCacheMap.get(stateType).refresh(accessor);
       }
-      Set<String> previousCachedStateTypes = _customizedViewCacheMap.keySet();
-      previousCachedStateTypes.removeAll(newStateTypes);
+      Set<String> previousCachedStateTypes = new HashSet<>(_customizedViewCacheMap.keySet());
+      previousCachedStateTypes.removeAll(_aggregationEnabledTypes);
       logger.info("Remove customizedView for state: " + previousCachedStateTypes);
       removeCustomizedViewTypes(new ArrayList<>(previousCachedStateTypes));
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java]: The previous change is a good catch. But I don't think we need this converting if we just make removeCustomizedViewTypes receives a set.

File: helix-core/src/main/java/org/apache/helix/controller/stages/CustomizedViewAggregationStage.java [hunk 7]
```diff
@@ -119,6 +117,7 @@ public void execute(final ClusterEvent event) throws Exception {
               "Failed to calculate customized view for resource " + resource.getResourceName(), ex);
         }
       }
+      updatedCustomizedViews.clear();
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/stages/CustomizedViewAggregationStage.java]: Why? This list is a local variable, right?

File: helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java (hunk 5 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java]: In refresh logic, the refresh of customized state config (aggregationEnableTypes) happened before customized view. If we read stale aggregationEnabledTypes, it means that the event of customized state config change has not been processed yet. And it should be fine we do not remove stale types, right?

File: helix-core/src/main/java/org/apache/helix/controller/stages/CustomizedViewAggregationStage.java (hunk 7 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/controller/stages/CustomizedViewAggregationStage.java]: It's a variable defined outside state type for loop. If we do not clear it, it actually contains resource -> customized view mapping for all state types. When we update ZK, there's problem as we are updating state by state. 
But I actually found this change is ugly, so I just moved the list inside the loop.

File: helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java (hunk 5 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java]: I think I misread. This is for the in-memory cache not for the znodes.

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java (hunk 4 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: Did you mean to remove this check or not?

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java [hunk 8]
```diff
@@ -139,7 +139,9 @@
   final AtomicReference<Map<String, LiveInstance>> _lastSeenInstances;
   final AtomicReference<Map<String, LiveInstance>> _lastSeenSessions;
 
-  final AtomicReference<Set<String>> _lastSeenCustomizedStateTypes;
+  // map that stores the mapping between instance and the customized state types available on that
+  //instance
+  final AtomicReference<Map<String, Set<String>>> _lastSeenCustomizedStateTypesMap;
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: nit, since I am confused again while reviewing the code related to this field, shall we rename it to _lastSeenCustomizedStateTypesRef Or _lastSeenCustomizedStateTypesMapRef?

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java [hunk 9]
```diff
@@ -828,12 +830,17 @@ public void onCustomizedStateRootChange(String instanceName, List<String> custom
...
+          .containsKey(instanceName)) {
+        lastSeenCustomizedStateTypes = lastSeenCustomizedStateTypesMap.get(instanceName);
+      }
       for (String customizedState : customizedStateTypes) {
         try {
-          if (lastSeenCustomizedStateTypes == null || !lastSeenCustomizedStateTypes
-              .contains(customizedState)) {
+          if (!lastSeenCustomizedStateTypes.contains(customizedState)) {
             manager.addCustomizedStateChangeListener(this, instanceName, customizedState);
             logger.info(
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: nit, if addCustomizedStateChangeListener already logs inside, we don't need this info log. Please confirm.

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java [hunk 10]
```diff
@@ -828,12 +830,17 @@ public void onCustomizedStateRootChange(String instanceName, List<String> custom
     }
 
     // TODO: remove the synchronization here once we move this update into dataCache.
-    synchronized (_lastSeenCustomizedStateTypes) {
-      Set<String> lastSeenCustomizedStateTypes = _lastSeenCustomizedStateTypes.get();
+    synchronized (_lastSeenCustomizedStateTypesMap) {
+      Map<String, Set<String>> lastSeenCustomizedStateTypesMap =
+          _lastSeenCustomizedStateTypesMap.get();
+      Set<String> lastSeenCustomizedStateTypes = Collections.emptySet();
+      if (lastSeenCustomizedStateTypesMap != null && lastSeenCustomizedStateTypesMap
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: Might help to simplify the code.

Map<String, Set<String>> lastSeenCustomizedStateTypesMap =
          _lastSeenCustomizedStateTypesMap.get();
if (null == lastSeenCustomizedStateTypesMap) {
  lastSeenCustomizedStateTypesMap = new HashMap();
  // lazy init the AtomicReference
  _lastSeenCustomizedStateTypesMapRef.set(lastSeenCustomizedStateTypesMap);
}
if (!lastSeenCustomizedStateTypesMap.containsKey(instanceName)) {
  lastSeenCustomizedStateTypesMap.put(instanceName, Collections.emptySet());
}
Set<String> lastSeenCustomizedStateTypes = lastSeenCustomizedStateTypesMap.get(instanceName);

Then you don't need to check null later.

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java (hunk 9 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: In ZkHelixManager, there's log for addListener, but in the controller, all the operations of adding listeners have this log, so I think it's better to keep consistent.

Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: The only additional information we get by printing this log is manager.getInstanceName(), I guess. I don't have a strong preference. But do whatever the other similar code does might cause redundant code. Please only keep this log if you think the extra helixmanager name can help us to debug in the future.

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java [hunk 11]
```diff
@@ -851,7 +865,7 @@ public void onCustomizedStateRootChange(String instanceName, List<String> custom
         }
       }
 
-      _lastSeenCustomizedStateTypes.set(new HashSet<>(customizedStateTypes));
+      lastSeenCustomizedStateTypes = new HashSet<>(customizedStateTypes);
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: Are you trying to do the following logic?

// Update the last seen types cache
lastSeenCustomizedStateTypes.clear()
lastSeenCustomizedStateTypes.addAll(customizedStateTypes)

Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: Yeah, messed it up.

File: helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java (hunk 9 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java]: I think it can still provide some information, so keep it for now.

**Comment by zhangmeng916** [comment]
This PR is read to be merged, approved by @jiajunwang 
Final commit message:
Fix minor issues in customized view aggregation logic and add some more tests.
